### PR TITLE
feat(number-input): ✨  add numberinput component

### PR DESCRIPTION
**Pull request by navin-moorthy**
Adds NumberInput Component

### CODE REVIEW

File: src/number-input/__utils.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+const FLOATING_POINT_REGEX = /^[Ee0-9\+\-\.]$/;
+
+/**
+ * Determine if a character is a DOM floating point character
+ * @see https://www.w3.org/TR/2012/WD-html-markup-20120329/datatypes.html#common.data.float
+ */
+export function isFloatingPointNumericCharacter(character: string) {
+  return FLOATING_POINT_REGEX.test(character);
+}
```
Comment (anuraghazra) [inline src/number-input/__utils.ts:12]: Interesting!

**Comment by anuraghazra** [comment]
@navin-moorthy A behaviour of native input which i'm seeing missing here is that we can focus on the input and use the mousewheel to increase/decrease the value.

File: src/number-input/__utils.ts (hunk 1 above)
Comment (navin-moorthy) [reply src/number-input/__utils.ts:12]: Quite nice. right. 
Only allowing the numbers to be typed and onChange.

File: src/number-input/NumberInputDecrementButton.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  Pick<Partial<NumberInputStateReturn>, "keepWithinRange"> &
+  Pick<
+    NumberInputStateReturn,
+    "focusInput" | "increment" | "decrement" | "isAtMin" | "spinner"
+  >;
+
+export type NumberInputDecrementButtonHTMLProps = ButtonHTMLProps;
+
+export type NumberInputDecrementButtonProps = NumberInputDecrementButtonOptions &
+  NumberInputDecrementButtonHTMLProps;
```
Comment (anuraghazra) [inline src/number-input/NumberInputDecrementButton.ts:19]: Unused Type

File: src/number-input/NumberInputState.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
+import { useCounter, UseCounterProps } from "@chakra-ui/counter";
+import { useBoolean } from "@chakra-ui/hooks";
+import { focus, minSafeInteger, maxSafeInteger } from "@chakra-ui/utils";
+import React, { useCallback, useRef } from "react";
```
Comment (anuraghazra) [inline src/number-input/NumberInputState.ts:4]: React isn't used, let's remove it.

**Comment by anuraghazra** [comment]
These two components are identical, i think we should refactor them.

`NumberInputDecrementButton` & `NumberInputIncrementButton`

File: src/number-input/NumberInputDecrementButton.ts (hunk 2 above)
Comment (navin-moorthy) [reply src/number-input/NumberInputDecrementButton.ts:19]: I am doing it with other fixes !!
Expect a commit soon.

**Comment by navin-moorthy** [comment]
> These two components are identical, i think we should refactor them.
> 
> `NumberInputDecrementButton` & `NumberInputIncrementButton`

Yup, I will create hook extracting the common behaviour !!

File: src/number-input/NumberInputDecrementButton.ts (hunk 2 above)
Comment (navin-moorthy) [reply src/number-input/NumberInputDecrementButton.ts:19]: Hey, just realised. This is not an unused Type.

This is used when we use this component as hooks, You can visit any of the reakit components and will see this props exported.

Useful to type the hook's props for the users who use it IMO.
Not really used within the components.

Comment (anuraghazra) [reply src/number-input/NumberInputDecrementButton.ts:19]: Ohhhh i see, okay okay got it!

**Comment by navin-moorthy** [comment]
> > These two components are identical, i think we should refactor them.
> > `NumberInputDecrementButton` & `NumberInputIncrementButton`
> 
> Yup, I will create hook extracting the common behaviour !!

Created the hook, just have a look !!

**Comment by anuraghazra** [comment]
Hey navin how about creating a HOC hook for reusing the logic? this way the code is less & types are too.

```ts
import { createHook } from "reakit-system";
import { ButtonHTMLProps, ButtonOptions, useButton } from "reakit/Button";

import { useSpinButton } from "./__utils";
import { NumberInputStateReturn } from "./NumberInputState";
import { NUMBERINPUT_DECREMENTBUTTON_KEYS } from "./__keys";

export type NumberInputButtonOptions = ButtonOptions &
  Pick<Partial<NumberInputStateReturn>, "keepWithinRange"> &
  Pick<
    NumberInputStateReturn,
    "focusInput" | "increment" | "decrement" | "isAtMin" | "isAtMax" | "spinner"
  >;

export const createNumberInputHook = (type: "increment" | "decrement") => {
  return createHook<NumberInputButtonOptions, ButtonHTMLProps>({
    name: "NumberInputDecrementButton",
    compose: useButton,
    keys: NUMBERINPUT_DECREMENTBUTTON_KEYS,

    useProps(options, htmlProps) {
      return useSpinButton(options, htmlProps, type);
    },
  });
};
```

NumberInputDecrementButton

```ts
import { createComponent } from "reakit-system";
import { createNumberInputHook } from "./createNumberInputHook";

export const NumberInputDecrementButton = createComponent({
  as: "button",
  memo: true,
  useHook: createNumberInputHook("decrement"),
});
```

Increment button

```js
import { createComponent } from "reakit-system";
import { createNumberInputHook } from "./createNumberInputHook";

export const NumberInputIncrementButton = createComponent({
  as: "button",
  memo: true,
  useHook: createNumberInputHook("increment"),
});

```


I have to code, let me know if you want me to just push it

**Comment by navin-moorthy** [comment]
> Hey @navin-moorthy how about creating a decorator function for reusing the logic? this way the code is less & types are too.
> 
> ```ts
> import { createHook } from "reakit-system";
> import { ButtonHTMLProps, ButtonOptions, useButton } from "reakit/Button";
> 
> import { useSpinButton } from "./__utils";
> import { NumberInputStateReturn } from "./NumberInputState";
> import { NUMBERINPUT_BUTTON_KEYS } from "./__keys";
> 
> export type NumberInputButtonOptions = ButtonOptions &
>   Pick<Partial<NumberInputStateReturn>, "keepWithinRange"> &
>   Pick<
>     NumberInputStateReturn,
>     "focusInput" | "increment" | "decrement" | "isAtMin" | "isAtMax" | "spinner"
>   >;
> 
> export const createNumberInputHook = (type: "increment" | "decrement") => {
>   return createHook<NumberInputButtonOptions, ButtonHTMLProps>({
>     name: `NumberInput`,
>     compose: useButton,
>     keys: NUMBERINPUT_BUTTON_KEYS,
> 
>     useProps(options, htmlProps) {
>       return useSpinButton(options, htmlProps, type);
>     },
>   });
> };
> ```
> 
> NumberInputDecrementButton
> 
> ```ts
> import { createComponent } from "reakit-system";
> import { createNumberInputHook } from "./createNumberInputHook";
> 
> export const NumberInputDecrementButton = createComponent({
>   as: "button",
>   memo: true,
>   useHook: createNumberInputHook("decrement"),
> });
> ```
> 
> Increment button
> 
> ```js
> import { createComponent } from "reakit-system";
> import { createNumberInputHook } from "./createNumberInputHook";
> 
> export const NumberInputIncrementButton = createComponent({
>   as: "button",
>   memo: true,
>   useHook: createNumberInputHook("increment"),
> });
> ```
> 
> I have the code, let me know if you want me to just push it

Yup, you can push it seems interesting ❤️ 
One more level abstraction !!

**Review by anuraghazra — APPROVED** [review]

**Review by anuraghazra — APPROVED** [review]

**Pull request merged**